File: SAS/OTB/OTB/src/nl/astron/lofar/sas/otbcomponents/bbs/stepmanagement/BBSStep.java
/*
 * BBSStep.java
 *
 *  Copyright (C) 2002-2007
 *  ASTRON (Netherlands Foundation for Research in Astronomy)
 *  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

package nl.astron.lofar.sas.otbcomponents.bbs.stepmanagement;

import java.util.HashMap;
import java.util.Vector;
import nl.astron.lofar.sas.otb.jotdb2.jOTDBnode;

/**
 * @version $Id$
 * @created July 25, 2006, 10:08 AM
 * @author pompert
 */
public class BBSStep{
    
    //Possible parent step
    private BBSStep parentStep;
    //Contained substeps
    private Vector<BBSStep> childSteps;
    //pointer to the Step Container in OTDB
    private jOTDBnode stepContainerNode = null;
    //Step Name
    private String name;
    //Step Baseline Selection
    private Vector<String> station1Selection = null;
    private Vector<String> station2Selection = null;
    //Step Sources
    private Vector<String> sources = null;
    private Vector<String> extraSources = null;
    //Step Instrument Model
    private Vector<String> instrumentModel = null;
    //Step Integration
    private double integrationFrequency = -1;
    private double integrationTime = -1;
    //Step Correlation
    private String correlationSelection = null;
    private Vector<String> correlationType = null;
    //Step Output Data Column
    private String outputDataColumn = null;
    
    //TODO: Step Operation Types
    
    
    /** Creates a new instance of BBSStep */
    public BBSStep(String name) {
        this.name=name;
        childSteps = new Vector<BBSStep>();
        parentStep = null;
    }
    
    public String getName(){
        return name;
    }
    
    public void setName(String name){
        this.name = name;
    }
    
    public Vector<String> getStation1Selection(){
        if(parentStep!=null && station1Selection == null){
            return parentStep.getStation1Selection();
        }
        return station1Selection;
    }
    
    public void setStation1Selection(Vector<String> station1Selection){
        this.station1Selection = station1Selection;
    }
    public Vector<String> getStation2Selection(){
        if(parentStep!=null  && station2Selection == null){
            return parentStep.getStation2Selection();
        }
        return station2Selection;
    }
    
    public void setStation2Selection(Vector<String> station2Selection){
        this.station2Selection = station2Selection;
    }
    public Vector<String> getSources(){
        if(parentStep!=null && sources == null){
            return parentStep.getSources();
        }
        return sources;
    }
    
    public void setSources(Vector<String> sources){
        this.sources = sources;
    }
    
    public Vector<String> getExtraSources(){
        if(parentStep!=null && extraSources == null){
            return parentStep.getExtraSources();
        }
        return extraSources;
    }
    
    public void setExtraSources(Vector<String> extraSources){
        this.extraSources = extraSources;
    }
    
    public Vector<String> getInstrumentModel(){
        if(parentStep!=null && instrumentModel == null){
            return parentStep.getInstrumentModel();
        }
        return instrumentModel;
    }
    
    public void setInstrumentModel(Vector<String> instrumentModel){
        this.instrumentModel = instrumentModel;
    }
    
    public double getIntegrationFrequency(){
        if(parentStep!=null && integrationFrequency == -1){
            return parentStep.getIntegrationFrequency();
        }
        return integrationFrequency;
    }
    
    public void setIntegrationFrequency(double integrationFrequency){
        this.integrationFrequency = integrationFrequency;
    }
    
    public double getIntegrationTime(){
        if(parentStep!=null && integrationTime == -1){
            return parentStep.getIntegrationTime();
        }
        return integrationTime;
    }
    
    public void setIntegrationTime(double integrationTime){
        this.integrationTime = integrationTime;
    }
    
    public String getCorrelationSelection(){
        if(parentStep!=null && correlationSelection == null){
            return parentStep.getCorrelationSelection();
        }
        return correlationSelection;
    }
    
    public void setCorrelationSelection(String correlationSelection){
        this.correlationSelection = correlationSelection;
    }
    
    public Vector<String> getCorrelationType(){
        if(parentStep!=null && correlationType == null){
            return parentStep.getCorrelationType();
        }
        return correlationType;
    }
    
    public void setCorrelationType(Vector<String> correlationType){
        this.correlationType = correlationType;
    }
    
    public String getOutputDataColumn(){
        if(parentStep!=null && outputDataColumn == null){
            return parentStep.getOutputDataColumn();
        }
        return outputDataColumn;
    }
    
    public void setOutputDataColumn(String outputDataColumn){
        this.outputDataColumn = outputDataColumn;
    }
    
    public BBSStep getParentStep(){
        return parentStep;
    }
    
    public void setParentStep(BBSStep parentStep){
        this.parentStep = parentStep;
    }
    
    public jOTDBnode getStepContainerPointer(){
        return stepContainerNode;
    }
    
    public void setStepContainerPointer(jOTDBnode stepContainerNode){
        this.stepContainerNode = stepContainerNode;
    }
    
    public boolean hasParentStep(){
        return parentStep != null;
    }
    
    public Vector<BBSStep> getChildSteps(){
        return childSteps;
    }
    
    public BBSStep getChildStep(String name) throws IllegalArgumentException{
        BBSStep returnStep = null;
        for(BBSStep aStep : childSteps){
            if(aStep.name.equals(name)){
                returnStep = aStep;
            }
        }
        if(returnStep==null){
            throw new IllegalArgumentException("No Child Step was found with name: "+name);
        }
        return returnStep;
    }
    
    public BBSStep getChildStepAtIndex(int index) throws IllegalArgumentException{
        BBSStep returnStep = null;
        
        if(childSteps.elementAt(index)!=null){
            returnStep = childSteps.get(index);
        }else{
            throw new IllegalArgumentException("No Child Step was found at index: "+index);
        }
        return returnStep;
    }
    public void addChildStep(BBSStep childStep){
        childStep.setParentStep(this);
        childSteps.add(childStep);
    }
    public void setChildStepAtIndex(BBSStep childStep, int index){
        childStep.setParentStep(this);
        childSteps.add(index,childStep);
    }
    public void removeChildStepAtIndex(int index){
        childSteps.get(index).setParentStep(null);
        childSteps.remove(index);
        childSteps.trimToSize();
    }
    
    public void removeChildStep(BBSStep childStep){
        if(childSteps.contains(childStep)){
            childSteps.remove(childStep);
            childSteps.trimToSize();
            childStep.setParentStep(null);            
        }        
    }
    
    public boolean hasChildSteps(){
        return childSteps.size()>0;
    }
    
    public void finalize(){
        if(parentStep!=null){
            parentStep.removeChildStep(this);
        }
    }    
}

File: branches/BBSKernel-Bug1519-task-branch/tags/CEP-BBSControl-Bug956-task-root/CEP/BB/BBSControl/src/BBSStep.cc
//#  BBSStep.cc: 
//#
//#  Copyright (C) 2002-2007
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#include <lofar_config.h>

#include <BBSControl/BBSStep.h>
#include <BBSControl/BBSCorrectStep.h>
#include <BBSControl/BBSPredictStep.h>
#include <BBSControl/BBSRefitStep.h>
#include <BBSControl/BBSShiftStep.h>
#include <BBSControl/BBSSolveStep.h>
#include <BBSControl/BBSSubtractStep.h>
#include <BBSControl/BBSMultiStep.h>
#include <BBSControl/Exceptions.h>
#include <Common/StreamUtil.h>
#include <APS/ParameterSet.h>
#include <APS/Exceptions.h>
#include <Common/LofarLogger.h>

namespace LOFAR
{
  using ACC::APS::ParameterSet;
  using ACC::APS::APSException;

  namespace BBS
  {
    using LOFAR::operator<<;

    //##--------   P u b l i c   m e t h o d s   --------##//

    BBSStep::~BBSStep()
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");
    }


    string BBSStep::fullName() const
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");
      string name;
      if (itsParent) name = itsParent->fullName() + ".";
      name += itsName;
      return name;
    }


    vector<const BBSStep*> BBSStep::getAllSteps() const
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");
      vector<const BBSStep*> steps;
      doGetAllSteps(steps);
      return steps;
    }


    BBSStep* BBSStep::create(const string& name,
			     const ParameterSet& parset,
			     const BBSStep* parent)
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");

      // If \a parset contains a key <tt>Step.<em>name</em>.Steps</tt>, then
      // \a name is a BBSMultiStep, otherwise it is a SingleStep.
      if (parset.isDefined("Step." + name + ".Steps")) {
	LOG_TRACE_COND_STR(name << " is a MultiStep");
	return new BBSMultiStep(name, parset, parent);
      } else {
	LOG_TRACE_COND_STR(name << " is a SingleStep");
	// We'll have to figure out what kind of SingleStep we must
	// create. The key "Operation" contains this information.
        try {
          string oper = 
            toUpper(parset.getString("Step." + name + ".Operation"));
          LOG_TRACE_COND_STR("Creating a " << oper << " step ...");
          if      (oper == "SOLVE")
            return new BBSSolveStep(name, parset, parent);
          else if (oper == "SUBTRACT")
            return new BBSSubtractStep(name, parset, parent);
          else if (oper == "CORRECT")
            return new BBSCorrectStep(name, parset, parent);
          else if (oper == "PREDICT")
            return new BBSPredictStep(name, parset, parent);
          else if (oper == "SHIFT")
            return new BBSShiftStep(name, parset, parent);
          else if (oper == "REFIT")
            return new BBSRefitStep(name, parset, parent);
          else THROW (BBSControlException, "Operation \"" << oper << 
                      "\" is not a valid Step operation");
        } catch (APSException& e) {
          THROW (BBSControlException, e.what());
        }
      }
    }


    //##--------   P r o t e c t e d   m e t h o d s   --------##//

    BBSStep::BBSStep(const string& name, 
		     const ParameterSet& parset,
		     const BBSStep* parent)
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");

      // Copy the data members from the parent, if present, so that they have
      // sensible default values.
      if (parent) *this = *parent;

      // We must reset these values, because they were overwritten by the
      // previous data copy.
      itsName = name;
      itsParent = parent;

      // Overrride default values for data members of the current BBSStep, if
      // they're specified in \a parset.
      read(parset.makeSubset("Step." + name + "."));

    }


    void BBSStep::write(ParameterSet& ps) const
    {
      LOG_TRACE_LIFETIME_STR(TRACE_LEVEL_COND, "Step." << itsName);
      ostringstream oss;
      oss << "Baselines.Station1 = "    << itsBaselines.station1    << endl
          << "Baselines.Station2 = "    << itsBaselines.station2    << endl
          << "Correlation.Selection = " << itsCorrelation.selection << endl
          << "Correlation.Type = "      << itsCorrelation.type      << endl
          << "Integration.Freq = "      << itsIntegration.deltaFreq << endl
          << "Integration.Time = "      << itsIntegration.deltaTime << endl
          << "Sources = "               << itsSources               << endl
          << "ExtraSources = "          << itsExtraSources          << endl
          << "InstrumentModel = "       << itsInstrumentModels      << endl;
      ps.adoptBuffer(oss.str(), "Step." + itsName + ".");
      LOG_TRACE_VAR_STR("\nContents of ParameterSet ps:\n" << ps);
    }


    void BBSStep::read(const ParameterSet& ps)
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");

      // If defined, get the baseline selection for this step.
      try {
	itsBaselines.station1 = ps.getStringVector("Baselines.Station1");
	itsBaselines.station2 = ps.getStringVector("Baselines.Station2");
      } catch (APSException&) {}

      // If defined, get the correlation selection (ALL, AUTO, or CROSS), and
      // type (e.g., ["XX", "XY", "YX", "YY"]
      try {
	itsCorrelation.selection = ps.getString("Correlation.Selection");
	itsCorrelation.type = ps.getStringVector("Correlation.Type");
      } catch (APSException&) {}

      // If defined, get the integration intervals in frequency (Hz) and
      // time (s).
      try {
	itsIntegration.deltaFreq = ps.getDouble("Integration.Freq");
	itsIntegration.deltaTime = ps.getDouble("Integration.Time");
      } catch (APSException&) {}

      // If defined, get the sources for the current patch.
      try {
	itsSources = ps.getStringVector("Sources");
      } catch (APSException&) {}

      // If defined, get the extra source, outside the current patch.
      try {
	itsExtraSources = ps.getStringVector("ExtraSources");
      } catch (APSException&) {}

      // If defined, get the instrument model(s) used.
      try {
	itsInstrumentModels = ps.getStringVector("InstrumentModel");
      } catch (APSException&) {}
    }


    void BBSStep::print(ostream& os) const
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");
      os << "Step: " << itsName;
      Indent id;  // add an extra indentation level
      os << endl << indent << "Full name: " << fullName()
	 << endl << indent << itsBaselines
	 << endl << indent << itsCorrelation
	 << endl << indent << itsIntegration
	 << endl << indent << "Sources: " << itsSources
	 << endl << indent << "Extra sources: " << itsExtraSources
	 << endl << indent << "Instrument models: " << itsInstrumentModels;
    }


    //##--------   P r i v a t e   m e t h o d s   --------##//

    void BBSStep::doGetAllSteps(vector<const BBSStep*>& steps) const
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");
      steps.push_back(this);
    }
    

    //##--------   G l o b a l   m e t h o d s   --------##//

    ostream& operator<<(ostream& os, const BBSStep& bs)
    {
      LOG_TRACE_LIFETIME(TRACE_LEVEL_COND, "");
      bs.print(os);
      return os;
    }


  } // namespace BBS

} // namespace LOFAR
